core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -254,15 +254,20 @@ private[spark] class TaskSchedulerImpl(
     // that stage when an active TaskSetManager succeed.
     stageIdToFinishedPartitions.getOrElseUpdate(taskSet.stageId, new BitSet)
     val tsm = new TaskSetManager(this, taskSet, maxTaskFailures, blacklistTrackerOpt)
-    // TaskSet got submitted by DAGScheduler may have some already completed
-    // tasks since DAGScheduler does not always know all the tasks that have
-    // been completed by other tasksets when completing a stage, so we mark
-    // those tasks as finished here to avoid launching duplicate tasks, while
-    // holding the TaskSchedulerImpl lock.
-    // See SPARK-25250 and `markPartitionCompletedInAllTaskSets()`
-    stageIdToFinishedPartitions.get(taskSet.stageId).foreach {
-      finishedPartitions => finishedPartitions.foreach(tsm.markPartitionCompleted(_, None))
+    // If this TaskSet belongs to a retrying indeterminate stage, we should create
+    // a total new TaskSetManager.
+    if (!taskSet.properties.containsKey(SparkContext.IS_INDETERMINATE_STAGE)) {
```

Comment:
I think we should add a new parameter in `TaskSet`.

Reply:
Yeah, make sense. https://github.com/apache/spark/pull/24110#discussion_r272801145